aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu

```diff
@@ -0,0 +1,155 @@
+  TORCH_CHECK(
+      canCast(commonDtype, output.scalar_type()),
+      "Can't convert result type ",
+      commonDtype,
+      " to output ",
+      output.scalar_type(),
+      " in add operation");
+
+  Tensor src_values = src.values().to(commonDtype);
+  Tensor src_crow_indices = src.crow_indices();
+  Tensor src_col_indices = src.col_indices();
+
+  output.resize_as_(dense);
+  Tensor resultBuffer = output;
+  Tensor valuesBuffer = src_values.to(commonDtype);
+
+  if (output.scalar_type() != commonDtype) {
+    resultBuffer = dense.to(commonDtype);
+  } else if (!is_same_tensor(output, dense)) {
+    resultBuffer.copy_(dense);
+  }
+  AT_DISPATCH_ALL_TYPES(
+      commonDtype,
+      "add_out_op2_sparse_csr",
+      [&src_values, &output, &alpha, &src_crow_indices, &src_col_indices]() {
+        AT_DISPATCH_INDEX_TYPES(
+            src_crow_indices.scalar_type(),
+            "csr_add_out_crow_indices",
+              [&src_values, &output, &alpha, &src_crow_indices, &src_col_indices]() {
+                scalar_t* values_accessor = src_values.data_ptr<scalar_t>();
+                scalar_t* out_ptr = output.data_ptr<scalar_t>();
+                scalar_t cast_value = alpha.to<scalar_t>();
+
+                index_t* crow_indices_accessor = src_crow_indices.data_ptr<index_t>();
+                index_t* col_indices_accessor = src_col_indices.data_ptr<index_t>();
+                int64_t out_storage_offset = output.storage_offset();
+
+                auto out_strides = output.strides();
+                int64_t out_strides0 = out_strides[0];
+                int64_t out_strides1 = out_strides[1];
+
+                cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+                auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+                auto policy = thrust::cuda::par(allocator).on(stream);
+
+               thrust::for_each(
+                    policy,
+                    thrust::make_counting_iterator(int64_t(0)),
+                    thrust::make_counting_iterator(int64_t(src_crow_indices.size(0) - 1)),
+                    [values_accessor,
+                    crow_indices_accessor,
+                    col_indices_accessor,
+                    out_ptr,
+                    out_storage_offset,
+                    out_strides0,
+                    cast_value,
+                    out_strides1
+                    ]__device__(int64_t irow) {
+                        int32_t start_index = crow_indices_accessor[irow];
+                        int32_t end_index = crow_indices_accessor[irow + 1];
+
+                        for (int i = start_index; i < end_index; ++i) {
+                            auto icol = col_indices_accessor[i];
+                            auto index = out_storage_offset + irow * out_strides0 + icol * out_strides1;
+                            out_ptr[index] += cast_value * values_accessor[i];
+                        }
```

Comment:
So this got me thinking. COO algorithms are basically CSR algorithms when the COO tensors are coalesced, right? So should we be setting up some code sharing between these two cases (COO coalesced, CSR)?

I went and looked at the COO algorithms; in particular mul (since add takes advantage of the fact that uncoalesced output is OK) and saw this (both sparse inputs were coalesced before this):

```
  Tensor resultNnz = at::empty({1}, CUDA(kLong));
  AT_DISPATCH_ALL_TYPES_AND(
    at::ScalarType::Half, commonDtype, "mul_out_sparse_cuda", [&] {
        apply::valueSparseIntersectionKernel<TensorMulOp<scalar_t>, uint64_t, scalar_t>
          <<<grid, block, 0, stream>>>(
            TensorMulOp<scalar_t>(),
            I_INFO(r_indices_), I_INFO(t_indices_), I_INFO(s_indices_),
            V_INFO(r_values_), V_INFO(t_values_), V_INFO(s_values_),
            static_cast<uint64_t>(t_nnz), static_cast<uint64_t>(s_nnz));
        C10_CUDA_KERNEL_LAUNCH_CHECK();
            
        apply::indexSparseIntersectionKernel<uint64_t, scalar_t>
          <<<1, 1, 0, stream>>>(
            I_INFO(r_indices_), I_INFO(t_indices_), I_INFO(s_indices_),
            // reinterpret_cast shenanigans, because we don't actually have
            // unsigned tensors...
            static_cast<uint64_t>(t_nnz), static_cast<uint64_t>(s_nnz), reinterpret_cast<uint64_t*>(resultNnz.data_ptr()));
        C10_CUDA_KERNEL_LAUNCH_CHECK();
      });   
```

aka we even are launching separate kernels for the values and indices. 

The intersection kernel doesn't look very good though:

```
template <typename Op, typename IndexType, typename Real>
#if __CUDA_ARCH__ >= 350 || defined __HIP_PLATFORM_HCC__
C10_LAUNCH_BOUNDS_2(cuda::getApplyBlockSize(), cuda::getApplyBlocksPerSM())
#endif
__global__ void valueSparseIntersectionKernel(
    Op op,
    TensorInfo<indexT, IndexType> r_indices,
    TensorInfo<indexT, IndexType> t_indices,
    TensorInfo<indexT, IndexType> s_indices,
    TensorInfo<Real, IndexType> r_values,
    TensorInfo<Real, IndexType> t_values,
    TensorInfo<Real, IndexType> s_values,
    const IndexType t_nnz, const IndexType s_nnz) {
  IndexType t_indskip = t_indices.strides[0];
  IndexType s_indskip = s_indices.strides[0];
  int64_t match, d;
  int64_t nDimI = r_indices.sizes[0];
  IndexType valueSize = r_values.strides[0];
  IndexType r_i = 0, t_i = 0, s_i = 0;
  while (t_i < t_nnz && s_i < s_nnz) {
    match = 1;
    for (d = 0; d < nDimI; d++) {
      if (t_indices.data[d * t_indskip + t_i] < s_indices.data[d * s_indskip + s_i]) {
        t_i++;
        match = 0;
        break;
      }
      if (t_indices.data[d * t_indskip + t_i] > s_indices.data[d * s_indskip + s_i]) {
        s_i++;
        match = 0;
        break;
      }
    }
    if (!match) continue;
    applyOp3(op, valueSize, r_values, r_i++, t_values, t_i++, s_values, s_i++);
  }
}
```

so maybe we should replace them with this thrust-based implementation here. Perhaps cc @ngimel

Reply:
You know what, never mind, you need crow indices so doesn't seem like any algorithm sharing can happen

Reply:
We do need to improve sparse multiplication, it's comically bad https://www.internalfb.com/phabricator/paste/view/P140370711. However, due to build and interop issues we are sunsetting thrust usage in the codebase and replacing it with cub wherever possible.